### PR TITLE
Feature/002 view job spec

### PR DESCRIPTION
**Pull request by Fluffyguy611**
# README - HOW TO TEST
1. Run purgeDataBase.sh
2. Run createDataBase.sh, populateDataBase.sh
3. Paste this into MySQLWorkBench:
```
UPDATE JobRole
SET SharePointLink = 'https://kainossoftwareltd.sharepoint.com/people/Job%20Specifications/Forms/AllItems.aspx?id=%2Fpeople%2FJob%20Specifications%2FEngineering%2FJob%20Profile%20%2D%20Principal%20Architect%20%28Principal%29%2Epdf&parent=%2Fpeople%2FJob%20Specifications%2FEngineering&p=true&ga=1'
WHERE Name = 'Principal';
```
#### needed for testing the SharepointLink is working, "Principal" should be under ID=2

4. RUN both backend and frontend app

##TEST SCENARIOS:
-missing role
-confirm link to sharepoint is working (check the url)
-confirm contents of pdf (automate)

#PROOF OF WORKING

Screenshot 2023-09-28 at 11 41 28
Screenshot 2023-09-28 at 11 41 23

# How to use parent.html

1. Add on the top of website:

```
{% extends "./parent.html" %}
{% block content %}
```

2. Start html inside <main></main>

3. End html file with:

```
{% endblock %}
```

CODE EXAMPLE:
```
{% extends "./parent.html" %}
{% block content %}
    
    <main class="container mt-4">
        <section id="jobRole" class="mt-4">
           {% if jobRole and jobRole.name.length > 0 %}
           <table class="table">
            <thead>
                <tr>
                    <th scope="col">Role Name</th>
                    <th scope="col">Role Description</th>
                    <th scope="col">SharePoint Link for Role Specification</th>
                </tr>
                <tbody>
                    <tr>
                        <td>{{ jobRole.name }}</td>
                        <td>{{ jobRole.description }} </td>
                        <td><a href="{{ jobRole.sharePointLink }}">View</a></td>
                    </tr>
                </tbody>
            </thead>
        </table>
           {% else %}
                Could not find job Role.
           {% endif %}
        </section>
    </main>

{% endblock %}
```

### CODE REVIEW

File: product-backend/src/main/java/org/kainos/ea/db/JobRoleSingleViewDaoA.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+public class JobRoleSingleViewDaoA {
+
+    public JobRoleSingleViewA getJobRoleById(int id, Connection c) throws SQLException {
+        String getStatement = "SELECT * FROM `JobRole` WHERE `Id`=" + id + ";";
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/db/JobRoleSingleViewDaoA.java]: You are not using prepared statement correctly - missing ? parameters

File: product-backend/src/main/java/org/kainos/ea/cli/JobRoleSingleViewA.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
+package org.kainos.ea.cli;
+public class JobRoleSingleViewA {
+
+        private int Id;
+        private String Name;
+        private String Description;
+        private String SharePointLink;
+
+    public JobRoleSingleViewA(int id, String name, String description, String sharePointLink) {
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/cli/JobRoleSingleViewA.java]:
```suggestion
    public JobRole(int id, String name, String description, String sharePointLink) {
```

File: product-backend/src/main/java/org/kainos/ea/api/JobRoleSingleViewService.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import org.kainos.ea.cli.JobRoleSingleViewA;
+import org.kainos.ea.client.DatabaseConnectionExceptionA;
+import org.kainos.ea.client.FailedToGetJobRoleException;
+import org.kainos.ea.client.JobRoleDoesNotExistException;
+import org.kainos.ea.db.DatabaseConnectorA;
+import org.kainos.ea.db.JobRoleSingleViewDaoA;
+
+import java.sql.SQLException;
+
+public class JobRoleSingleViewService {
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/api/JobRoleSingleViewService.java]:
```suggestion
public class JobRoleService {
```

File: product-backend/src/main/java/org/kainos/ea/cli/JobRoleSingleViewA.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
+package org.kainos.ea.cli;
+public class JobRoleSingleViewA {
+
+        private int Id;
+        private String Name;
+        private String Description;
+        private String SharePointLink;
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/cli/JobRoleSingleViewA.java]:
```suggestion
        private int id;
        private String iame;
        private String description;
        private String sharePointLink;
```

File: product-backend/src/main/java/org/kainos/ea/controller/JobRoleSingleViewControllerA.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+
+    @GET
+    @Path("/jobRoles/{id}")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getJobRoleById(@PathParam("id") int id) {
+        try {
+            return Response.ok(jobRoleSingleViewService.getJobRoleSingleViewById(id)).build();
+        } catch ( FailedToGetJobRoleException e) {
+            System.err.println(e.getMessage());
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/controller/JobRoleSingleViewControllerA.java]: use logger, see my example controller

File: product-backend/src/main/java/org/kainos/ea/db/JobRoleSingleViewDaoA.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+                    rs.getInt("Id"),
+                    rs.getString("Name"),
+                    rs.getString("Description"),
+                    rs.getString("SharePointLink")
+            );
+
+        }
+
+        return null;
+    }
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/db/JobRoleSingleViewDaoA.java]: Use optionals
https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html
```suggestion
    public Optional<JobRole> getJobRoleById(int id, Connection c) throws SQLException {
        String getStatement = "SELECT * FROM `JobRole` WHERE `Id`=" + id + ";";
        PreparedStatement st = c.prepareStatement(getStatement);
        ResultSet rs = st.executeQuery();

        while (rs.next()){
            return Optional.of(new JobRoleSingleViewA(
                    rs.getInt("Id"),
                    rs.getString("Name"),
                    rs.getString("Description"),
                    rs.getString("SharePointLink"))
            );

        }

        return Optional.empty();
    }
```

Service usage:

```
 Optional<JobRole> jobRole = jobRoleSingleViewDao.getJobRoleById(id, databaseConnector.getConnection());
return jobRole.orElseThrow(() -> new JobRoleDoesNotExistException());
```

File: product-web-app/service/jobRoleSingleViewService.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import axios from 'axios';
+import JobRole from '../model/jobRoleSingleView.js';
+import logger from './logger.js';
+
+
+export default class JobRoleService {
+
+  async getJobRoleSpecification(id: number): Promise<JobRole> {
+    try {
+      const response = await axios.get('http://localhost:8080/api/jobRoles/' + id,);
```
Comment (ljercha) [inline product-web-app/service/jobRoleSingleViewService.ts]: Localhost is redudent, we set base url in app.ts
```
axios.defaults.baseURL = API_URL;
```

```suggestion
      const response = await axios.get(API.GET_JOB_ROLES(id));
```

File: product-web-app/common/constants.ts [hunk 8]
```diff
@@ -1,3 +1,6 @@
-const API_URL = `${process.env.API_URL}`;
+export const API_URL = `${process.env.API_URL}`;
 
-export default API_URL;
+export const API = {
+  PRODUCTS: '/api/jobRoles',
+  GET_JOBROLES: (id: number) => `/api/jobRoles/${id}`,
```
Comment (ljercha) [inline product-web-app/common/constants.ts]:
```suggestion
  GET_JOB_ROLE: (id: number) => `/api/jobRoles/${id}`,
```

File: product-web-app/views/view-single-jobRole.html [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css">
+</head>
+<body>
+    <header class="bg-dark text-light">
+        <div class="container">
+            <div class="row">
+                <div class="col-md-12">
+                    <h1>Job Role</h1>
+                </div>
+                {% if errormessage %}
```
Comment (ljercha) [inline product-web-app/views/view-single-jobRole.html]: not used

File: product-web-app/views/view-single-jobRole.html [hunk 10]
```diff
@@ -0,0 +1,48 @@
+<!DOCTYPE html>
+<html lang="en">
```
Comment (ljercha) [inline product-web-app/views/view-single-jobRole.html]: Can you create template for HTML/Header document
See: https://mozilla.github.io/nunjucks/templating.html

File: product-web-app/tsconfig.json [hunk 11]
```diff
@@ -26,7 +26,7 @@
     // "useDefineForClassFields": true,                  /* Emit ECMAScript-standard-compliant class fields. */
     // "moduleDetection": "auto",                        /* Control what method is used to detect module-format JS files. */
 
-    "moduleResolution": "Node16",
+    "moduleResolution": "node16",
```
Comment (ljercha) [inline product-web-app/tsconfig.json]: you shouldn't modify it

**Review by ljercha — COMMENTED** [review]
General note:
Remove the 'A' postfix from classes/services
Remember about the unit test for services (JobRoleService in TS and in Java)

**Review by BartlomiejMilewski — APPROVED** [review]

File: product-backend/src/main/java/org/kainos/ea/db/JobRoleDao.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+import org.kainos.ea.cli.JobRole;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+public class JobRoleDao {
+
+    public JobRole getJobRoleById(int id, Connection c) throws SQLException {
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/db/JobRoleDao.java]: Change to return optional - @BartlomiejMilewski already did it

File: product-web-app/common/constants.ts [hunk 13]
```diff
@@ -1,3 +1,6 @@
-const API_URL = `${process.env.API_URL}`;
+export const API_URL = `${process.env.API_URL}`;
 
-export default API_URL;
+export const API = {
+  JOB_ROLES: '/api/jobRoles/',
+  GET_JOB_ROLE: (id: number) => `api/jobRoles/${id}`,
```
Comment (ljercha) [inline product-web-app/common/constants.ts]: let's keep it lower case and dashes
```suggestion
  JOB_ROLES: '/api/job-roles/',
  GET_JOB_ROLE: (id: number) => `api/job-roles/${id}`,
```

File: product-backend/src/main/java/org/kainos/ea/api/JobRoleService.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+    public JobRoleService(DatabaseConnector databaseConnector, JobRoleDao jobRoleDao) {
+        this.databaseConnector = databaseConnector;
+        this.jobRoleDao = jobRoleDao;
+    }
+
+
+    public JobRole getJobRoleById(int id) throws JobRoleDoesNotExistException, FailedToGetJobRoleException {
+        try {
+            JobRole jobRoleSingleView = jobRoleDao.getJobRoleById(id, databaseConnector.getConnection());
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/api/JobRoleService.java]:
```suggestion
            JobRole jobRole = jobRoleDao.getJobRoleById(id, databaseConnector.getConnection());
```

File: product-backend/src/main/java/org/kainos/ea/controller/JobRoleController.java [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+@Path("/api")
+public class JobRoleController {
+    private final static Logger logger = LoggerFactory.getLogger(JobRoleService.class);
+
+    JobRoleService jobRoleService = new JobRoleService(new DatabaseConnector(), new JobRoleDao());
+
+
+    @GET
+    @Path("/jobRoles/{id}")
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/controller/JobRoleController.java]:
```suggestion
    @Path("/job-roles/{id}")
```

**Review by mtslech — APPROVED** [review]

File: product-web-app/service/jobRoleService.ts [hunk 16]
```diff
@@ -0,0 +1,19 @@
+import axios from 'axios';
+import JobRole from '../model/jobRole.js';
+import logger from './logger.js';
+import  { API } from '../common/constants.js'
+
+
+export default class JobRoleService {
```
Comment (ljercha) [inline product-web-app/service/jobRoleService.ts:6]: Missing unit test - see in my examples

**Review by ljercha — COMMENTED** [review]
almost ready for testing!

**Review by AgaGorlak — APPROVED** [review]

File: product-backend/src/main/java/org/kainos/ea/db/JobRoleDao.java [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+                    rs.getInt("Id"),
+                    rs.getString("Name"),
+                    rs.getString("Description"),
+                    rs.getString("SharePointLink")
+            ));
+
+        }
+
+        return null;
+    }
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/db/JobRoleDao.java]:
```suggestion
        return Optional.empty();
    }
```

**Review by ljercha — APPROVED** [review]
Just 1 comment expect that looks good

**Pull request merged**